# Advanced Data Analytics Task

## Task 1 

Data Cleaning Challenge on Sales Data. 

In [198]:
import pandas as pd

sales_data_2017 = pd.read_csv("Sales Transactions-2017.csv")
sales_data_2018 = pd.read_csv("Sales Transactions-2018.csv")
sales_data_2019 = pd.read_csv("Sales Transactions-2019.csv")

total_sales_data = pd.concat(objs = [sales_data_2017, sales_data_2018, sales_data_2019])
total_sales_data

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
0,1/4/2017,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00","3,380.00",NaN,"13,100.00"
1,1/4/2017,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
2,1/4/2017,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23,"11,500.00",NaN,"30,990.00"
3,1/4/2017,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
4,1/4/2017,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00","8,450.00",NaN,NaN
...,...,...,...,...,...,...,...,...,...
19171,10/10/2019,Sal:4935,K.SRIHARI,13*16 WHITE RK,400,16,"6,400.00",NaN,NaN
19172,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
19173,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
19174,NaN,Total,NaN,NaN,"99,284.90","175,381.65","2,203,649.50","20,680.00","2,189,014.50"


In [199]:
print("Original Shape = ", total_sales_data.shape)

Original Shape =  (111206, 9)


In [200]:
total_sales_data.isnull().sum()

Date               12591
Voucher            12557
Party                 40
Product            12591
Qty                12557
Rate               12558
Gross              12558
Disc              105609
Voucher Amount     83646
dtype: int64

In [201]:
total_sales_data.isna().sum()

Date               12591
Voucher            12557
Party                 40
Product            12591
Qty                12557
Rate               12558
Gross              12558
Disc              105609
Voucher Amount     83646
dtype: int64

In [202]:
total_sales_data.duplicated().sum()

np.int64(14582)

In [203]:
total_sales_data.drop_duplicates(inplace= True)
total_sales_data

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
0,1/4/2017,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00","3,380.00",NaN,"13,100.00"
1,1/4/2017,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
2,1/4/2017,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23,"11,500.00",NaN,"30,990.00"
3,1/4/2017,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
4,1/4/2017,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00","8,450.00",NaN,NaN
...,...,...,...,...,...,...,...,...,...
19169,10/10/2019,Sal:4935,K.SRIHARI,13*16 Bk(100)KRISHN,320,16,"5,120.00",NaN,NaN
19170,10/10/2019,Sal:4935,K.SRIHARI,10*12 RK,800,8.5,"6,800.00",NaN,NaN
19171,10/10/2019,Sal:4935,K.SRIHARI,13*16 WHITE RK,400,16,"6,400.00",NaN,NaN
19174,NaN,Total,NaN,NaN,"99,284.90","175,381.65","2,203,649.50","20,680.00","2,189,014.50"


In [204]:
print("After Duplicate Records Removed Shape = ", total_sales_data.shape)

After Duplicate Records Removed Shape =  (96624, 9)


#### Shape Details
Original Shape =  (111206, 9)

After Duplicate Records Removed Shape =  (96624, 9)

In [205]:
total_sales_data.dtypes

Date              str
Voucher           str
Party             str
Product           str
Qty               str
Rate              str
Gross             str
Disc              str
Voucher Amount    str
dtype: object

In [206]:
total_sales_data.Date

0          1/4/2017
1          1/4/2017
2          1/4/2017
3          1/4/2017
4          1/4/2017
            ...    
19169    10/10/2019
19170    10/10/2019
19171    10/10/2019
19174           NaN
19175           NaN
Name: Date, Length: 96624, dtype: str

In [207]:
total_sales_data.Date.isna().sum()

np.int64(204)

In [208]:
total_sales_data[total_sales_data.Date.isna()]

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
11,NaN,NaN,#NAME?,NaN,NaN,NaN,NaN,NaN,NaN
53,NaN,NaN,KR,NaN,NaN,NaN,NaN,NaN,NaN
77,NaN,NaN,IC CENTRE,NaN,NaN,NaN,NaN,NaN,NaN
79,NaN,NaN,LAKSHMIMEDICALS,NaN,NaN,NaN,NaN,NaN,NaN
104,NaN,NaN,LS VGIRI,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...
18033,NaN,NaN,RU,NaN,NaN,NaN,NaN,NaN,NaN
18347,NaN,Total,NaN,NaN,"274,408.60","674,780.19","7,168,515.80","69,050.00","7,119,450.80"
18601,NaN,NaN,RESH,NaN,NaN,NaN,NaN,NaN,NaN
19174,NaN,Total,NaN,NaN,"99,284.90","175,381.65","2,203,649.50","20,680.00","2,189,014.50"


In [209]:
total_sales_data.dropna(subset="Date", inplace= True)
total_sales_data.Date.isna().sum()

np.int64(0)

In [210]:
print("After Date nan Records Removed Shape = ",total_sales_data.shape)

After Date nan Records Removed Shape =  (96420, 9)


#### Shape Details After Date nan values removed
Original Shape =  (111206, 9)

After Duplicate Records Removed Shape =  (96624, 9)

After Date nan Records Removed Shape =  (96420, 9)


In [211]:
total_sales_data.Date.dtypes

<StringDtype(storage='python', na_value=nan)>

In [212]:
import numpy as np

total_sales_data.Date = total_sales_data.Date.str.strip()
total_sales_data.Date = total_sales_data.Date.replace("", np.nan)


In [213]:
total_sales_data.dropna(subset="Date", inplace=True)
total_sales_data

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
0,1/4/2017,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00","3,380.00",NaN,"13,100.00"
1,1/4/2017,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
2,1/4/2017,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23,"11,500.00",NaN,"30,990.00"
3,1/4/2017,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
4,1/4/2017,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00","8,450.00",NaN,NaN
...,...,...,...,...,...,...,...,...,...
19167,10/10/2019,Sal:4935,K.SRIHARI,16*20(100-W),140,26,"3,640.00",NaN,NaN
19168,10/10/2019,Sal:4935,K.SRIHARI,10*12 KRISHNA-BK(10,600,8.4,"5,040.00",NaN,NaN
19169,10/10/2019,Sal:4935,K.SRIHARI,13*16 Bk(100)KRISHN,320,16,"5,120.00",NaN,NaN
19170,10/10/2019,Sal:4935,K.SRIHARI,10*12 RK,800,8.5,"6,800.00",NaN,NaN


In [214]:
print("After Date empty Records Removed Shape = ",total_sales_data.shape)

After Date empty Records Removed Shape =  (95287, 9)


#### Shape Details After Date nan values removed
Original Shape =  (111206, 9)

After Duplicate Records Removed Shape =  (96624, 9)

After Date nan Records Removed Shape =  (96420, 9)

After Date empty Records Removed Shape =  (95287, 9)


In [215]:
total_sales_data.dtypes

Date              str
Voucher           str
Party             str
Product           str
Qty               str
Rate              str
Gross             str
Disc              str
Voucher Amount    str
dtype: object

In [216]:
total_sales_data["Date"] = pd.to_datetime(
    total_sales_data["Date"],
    dayfirst=True
)  
total_sales_data.dtypes

Date              datetime64[us]
Voucher                      str
Party                        str
Product                      str
Qty                          str
Rate                         str
Gross                        str
Disc                         str
Voucher Amount               str
dtype: object

In [217]:
total_sales_data.isna().sum()

Date                  0
Voucher               0
Party                 0
Product               0
Qty                   0
Rate                  1
Gross                 1
Disc              92781
Voucher Amount    70814
dtype: int64

In [218]:
total_sales_data[total_sales_data.Rate.isna()]

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
43740,2018-03-05,Sal:9266,CUSTOMER,TIRML-BL-KG-17*23,2,NaN,NaN,NaN,NaN


In [219]:
import numpy as np

total_sales_data["Rate"] = total_sales_data["Rate"].str.strip()
total_sales_data["Rate"] = total_sales_data["Rate"].replace("", np.nan)

total_sales_data.dropna(subset=["Rate"], inplace=True)

In [220]:
total_sales_data.Rate.isna().sum()

np.int64(0)

In [174]:
print("After Rate NAN Records Removed Shape = ",total_sales_data.shape)

After Rate NAN Records Removed Shape =  (95286, 9)


#### Shape Details After Date nan values removed
Original Shape =  (111206, 9)

After Duplicate Records Removed Shape =  (96624, 9)

After Date nan Records Removed Shape =  (96420, 9)

After Date empty Records Removed Shape =  (95287, 9)

After Rate NAN Records Removed Shape =  (95286, 9)


In [222]:
total_sales_data.isnull().sum()

Date                  0
Voucher               0
Party                 0
Product               0
Qty                   0
Rate                  0
Gross                 0
Disc              92780
Voucher Amount    70813
dtype: int64

In [223]:
total_sales_data.Voucher = total_sales_data.Voucher.str.replace("Sal:", "").astype(int)

In [224]:
total_sales_data.Voucher.head()

0    1
1    1
2    2
3    2
4    2
Name: Voucher, dtype: int64

In [225]:
total_sales_data.dtypes

Date              datetime64[us]
Voucher                    int64
Party                        str
Product                      str
Qty                          str
Rate                         str
Gross                        str
Disc                         str
Voucher Amount               str
dtype: object

In [226]:
total_sales_data.Qty = total_sales_data.Qty.str.replace(",", "").astype(float).astype(int)


In [227]:
total_sales_data.dtypes

Date              datetime64[us]
Voucher                    int64
Party                        str
Product                      str
Qty                        int64
Rate                         str
Gross                        str
Disc                         str
Voucher Amount               str
dtype: object

In [228]:
total_sales_data.Rate = total_sales_data.Rate.str.replace(",","").astype(float)

In [229]:
total_sales_data.dtypes

Date              datetime64[us]
Voucher                    int64
Party                        str
Product                      str
Qty                        int64
Rate                     float64
Gross                        str
Disc                         str
Voucher Amount               str
dtype: object

In [230]:
total_sales_data.shape

(95286, 9)

## Task 2

## Merge Function 1 Example 

In [235]:
import pandas as pd

students = pd.DataFrame({
    "Student_ID": [101, 102, 103, 104],
    "Name": ["Alice", "Bob", "Charlie", "David"]
})

marks = pd.DataFrame({
    "Student_ID": [101, 102, 104, 105],
    "Marks": [85, 90, 78, 88]
})

result = pd.merge(students, marks, on="Student_ID", how="inner")
result

,Student_ID,Name,Marks
0,101,Alice,85
1,102,Bob,90
2,104,David,78


In [236]:
result = pd.merge(students, marks, on="Student_ID", how="left")
result

,Student_ID,Name,Marks
0,101,Alice,85.0
1,102,Bob,90.0
2,103,Charlie,NaN
3,104,David,78.0


In [237]:
result = pd.merge(students, marks, on="Student_ID", how="right")
result

,Student_ID,Name,Marks
0,101,Alice,85
1,102,Bob,90
2,104,David,78
3,105,NaN,88


In [238]:
result = pd.merge(students, marks, on="Student_ID", how="outer")
result

,Student_ID,Name,Marks
0,101,Alice,85.0
1,102,Bob,90.0
2,103,Charlie,NaN
3,104,David,78.0
4,105,NaN,88.0


# Task 3 

## List Comprehension 2 examples 

In [250]:
# Example 1 salary greater than 50,000

employees = [
    {"name": "Raj", "salary": 45000},
    {"name": "Sam", "salary": 60000},
    {"name": "Ram", "salary": 75000},
    {"name": "Hari", "salary": 40000}
]

emp_name = [emp["name"] for emp in employees if emp["salary"] > 50000]
emp_name

['Sam', 'Ram']

In [279]:
# Example 2
list1 = [1, 2, 3]
list2 = [10, 20]

result = [(i, j) for i in list1 for j in list2 if j>10]
result

[(1, 20), (2, 20), (3, 20)]

# Task 4

## Lambda Function 2 examples with Map, Filter, Reduce

### Map

In [264]:
# Example 1: Square every number
numbers = [1,2, 3, 4, 5]

result = list(map(lambda x: x**2, numbers))
result

[1, 4, 9, 16, 25]

In [266]:
# Example 2: Convert names to uppercase
names = ["raj", "sam", "ram"]

result = list(map(lambda x: x.upper(), names))
result

['RAJ', 'SAM', 'RAM']

### Filter

In [272]:
# Example 1: Get even numbers
numbers = [1,2,3,4,5,6,7,8]

result = list(filter(lambda x: x%2==0, numbers))
print(result)

[2, 4, 6, 8]


In [273]:
# Example 2: Students with marks >= 50

marks = [45, 67, 30, 89, 55]
result = list(filter(lambda x: x>=50, marks))
print(result)

[67, 89, 55]


### Reduce

In [274]:
# Example 1: Sum of numbers

from functools import reduce
numbers = [1,2,3,4,5]
result = reduce(lambda x,y: x+y, numbers)
result

15

In [277]:
# Example 2: Find the largest number
from functools import reduce

numbers = [12, 8, 45, 22, 30]
result = reduce(lambda x, y: x if x>y else y, numbers)
result

45